### ISSUE TITLE
Container patterns with variably named files

### ISSUE THREAD

**Issue by AndreaHri** [digital-preservation/droid#823]
Container signature identification is currently based on finding files within container with specific exact name syntax. However, this is not true for many formats which will have same file type with its container content, but the name of that file may vary. Please see the full discussion and recommendations by the community on https://github.com/digital-preservation/pronom/issues/10

**Comment by steve-daly** [comment]
I would suggest allowing Glob patterns (e.g. ?, *) in Container Signature filenames.  To be honest, I don't know what happens currently if you try this.  I imagine we could add these patterns in PRONOM without any code changes there. 

For the DROID change, we'd just need to consider what we do with multiple matches. 

For example, if we are looking for a XML file with a root tag of <Autodesk Version="1.0"> inside an Autodesk RCP file, but where the XML filename will vary in each file, as it's a UUID. 

We could specify the XML filename as `*.xml` but what happens if there is more than one file matching that pattern?  Do we:

- Require all files matching the filename pattern to also match the subsequent criterion (e.g. that root tag)
- Keep trying files matching that pattern until we find one that matches the subsequent criterion, flagging as a good match for RCP when we find it
- Only check the first match, as we're only really supporting the wildcard to match a single file, as that's all we expect from the examples so far. 

These are probably equally easy to implement, we'd just want to consider our preferred behaviour.  All three would work for the examples we've seen before, but some are tighter than others in failing on unexpected variations.

**Comment by Dclipsham** [comment]
Be aware that container triggers include Zip files, and that most Zip files are not 'container formats' in the DROID-sense so having very liberal filename pattern matching may significantly increase the processing time for these files, and may also result in false positives

**Comment by steve-daly** [comment]
I may not be understanding this fully, but I'm thinking this relates to the `Path `element, for example:

```
<ContainerSignature Id="58000" ContainerType="ZIP">
  <Description>ISDOCX Information System Document (version 6.0.1)</Description>
    <Files>
      <File>
        <Path>manifest.xml</Path>
          <BinarySignatures>
            <InternalSignatureCollection>
              <InternalSignature ID="1">
                <ByteSequence Reference="BOFoffset"> 
                  <SubSequence Position="1" SubSeqMinOffset="50" SubSeqMaxOffset="100">
                    <Sequence>'isdoc.cz/namespace/2013/manifest'</Sequence>
                  </SubSequence>
                </ByteSequence>
             </InternalSignature>
           </InternalSignatureCollection>
         </BinarySignatures>
      </File>
   </Files>
</ContainerSignature>
```

where it says <Path>manifest.xml</Path> but where the equivalent of manifest.xml (e.g. a UUID suffix) doesn't have a consistent filename between file instances, we could instead say something like...

<Path>manifest_*.xml</Path>

I'm thinking DROID would initially trigger on identifying a file as a ZIP, and then see if there are any files within matching "manifest_*.xml" within the ZIP before looking within those files for the ByteSequence. 

Do you mean that if we had something less tight such as <Path>*</Path> then we'd be searching vastly more files within ZIPs looking for the ByteSequences.  

I would be hoping that manifest_*.xml would be equally common as manifest.xml.  I suspect I may be missing something here

**Comment by Dclipsham** [comment]
Yup so right now where we're being specific we're only really seeking a filepath from the zip root and if it exists then scanning the internals of the given file(s), so we're typically seeking the presence of a single file - possibly a handful (e.g. ESRI ArcScene Document where we're seeking 5 specific files), which keeps things targetted and reasonably efficient.

In the case where we're able to express any pattern like *.xml we'd be asking it to scan all instances of files with an .xml extension within every zip, which could be pretty wasteful if its just a large zip of XML files

So in the cases in https://github.com/digital-preservation/pronom/issues/10 where we've potentially got regex patterns we could express, like:
Thumbs.db: "256_(([a-z]|[0-9]){16})"
IWA: "slide-([0-9]{1,5}).iwa"
or possibly even:
RCP: "(([a-z]|[0-9]){8})-(([a-z]|[0-9]){4})-(([a-z]|[0-9]){4})-(([a-z]|[0-9]){4})-(([a-z]|[0-9]){12}).xml"
Then these might be specific enough to process efficiently, but if we just wanted to express something like:
"*.xml" -  I fear this would be too general.

So, I'm just suggesting such a feature needs careful handling, and might also benefit from being introduced beind a feature switch so users need to actuvely turn it on to make use of it...

**Comment by steve-daly** [comment]
Thanks @Dclipsham Maybe that makes a case for regex rather than glob for the filename matching, to be more specific.

**Comment by Dclipsham** [comment]
well regex was just the easiest way for me to express the repetition above - as Richard points out on the other thread, glob is potentially less complex and more consistent than regex in implementation for tools like DROID & Siegfried, but it would likely necessitate making your identification patterns far more verbose
